### PR TITLE
Better nullable primary key implementation

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Get rid of werid variant types in Field. Instead we hold variants inside Null.

Detailed description / Documentation draft:
.

### CODE REVIEW

File: src/Core/Types.h [hunk 1]
```diff
@@ -13,9 +13,23 @@ namespace DB
...
 /// Data types for representing elementary values from a database in RAM.
 
-struct Null {};
-struct NegativeInfinity {};
-struct PositiveInfinity {};
+/// Hold a null value for untyped calculation. It can also store infinities to handle nullable
+/// comparison which is used for nullable KeyCondition.
+struct Null
+{
+    bool is_negative_infinity = false;
```
Comment (alexey-milovidov) [inline src/Core/Types.h:20]: It's strange that it can both positive and negative infinity.
Maybe enum?

**Review by alexey-milovidov — APPROVED** [review]
Still has some weirdness, but overall change is good.

**Comment by amosbird** [comment]
fuzz crash query : `SELECT count(-1), count(NULL) = count() FROM (SELECT count(1.000100016593933) = NULL, count(65535), count(NULL) = count()) AS js1 ALL RIGHT JOIN (SELECT NULL, number AS s FROM numbers(1025)) AS js2 USING (s)`


```
<Fatal> BaseDaemon: ########################################
<Fatal> BaseDaemon: (version 21.10.1.7933, build id: 5DD1E302B6510800100F55204F3FF8648854F6FE) (from thread 315) (query_id: dac87a4a-edb3-45ce-892a-892fab2e53ee) Received signal Segmentation fault (11)
<Fatal> BaseDaemon: Address: 0x2f Access: read. Address not mapped to object.
<Fatal> BaseDaemon: Stack trace: 0x238c4685 0x233f5360 0x254eea91 0x250f1f13 0x250eca9b 0x250e77d3 0x250e6ac8 0x2510c8c9 0xbdb7b8c 0xbdbf274 0x7f262591a609 0x7f2625841293
<Fatal> BaseDaemon: 3.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/string:1444: std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__is_long() const
<Fatal> BaseDaemon: 3.2. inlined from ../contrib/libcxx/include/string:1902: basic_string
<Fatal> BaseDaemon: 3. ../src/Interpreters/join_common.cpp:507: DB::NotJoinedBlocks::NotJoinedBlocks(std::__1::unique_ptr<DB::NotJoinedBlocks::RightColumnsFiller, std::__1::default_delete<DB::NotJoinedBlocks::RightColumnsFiller> >, DB::Block const&, unsigned long, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > > const&) @ 0x238c4685 in /workspace/clickhouse
<Fatal> BaseDaemon: 4. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:0: DB::MergeJoin::getNonJoinedBlocks(DB::Block const&, unsigned long) const @ 0x233f5360 in /workspace/clickhouse
<Fatal> BaseDaemon: 5.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/type_traits:3935: std::__1::enable_if<(is_move_constructible<DB::NotJoinedBlocks*>::value) && (is_move_assignable<DB::NotJoinedBlocks*>::value), void>::type std::__1::swap<DB::NotJoinedBlocks*>(DB::NotJoinedBlocks*&, DB::NotJoinedBlocks*&)
<Fatal> BaseDaemon: 5.2. inlined from ../contrib/libcxx/include/memory:3299: std::__1::shared_ptr<DB::NotJoinedBlocks>::swap(std::__1::shared_ptr<DB::NotJoinedBlocks>&)
<Fatal> BaseDaemon: 5.3. inlined from ../contrib/libcxx/include/memory:3243: std::__1::shared_ptr<DB::NotJoinedBlocks>::operator=(std::__1::shared_ptr<DB::NotJoinedBlocks>&&)
<Fatal> BaseDaemon: 5. ../src/Processors/Transforms/JoiningTransform.cpp:124: DB::JoiningTransform::work() @ 0x254eea91 in /workspace/clickhouse
<Fatal> BaseDaemon: 6.1. inlined from ./obj-x86_64-linux-gnu/../src/Processors/Executors/PipelineExecutor.cpp:0: operator()
<Fatal> BaseDaemon: 6.2. inlined from ../contrib/libcxx/include/type_traits:3676: decltype(std::__1::forward<DB::PipelineExecutor::addJob(DB::ExecutingGraph::Node*)::$_0&>(fp)()) std::__1::__invoke<DB::PipelineExecutor::addJob(DB::ExecutingGraph::Node*)::$_0&>(DB::PipelineExecutor::addJob(DB::ExecutingGraph::Node*)::$_0&)
<Fatal> BaseDaemon: 6.3. inlined from ../contrib/libcxx/include/__functional_base:348: void std::__1::__invoke_void_return_wrapper<void>::__call<DB::PipelineExecutor::addJob(DB::ExecutingGraph::Node*)::$_0&>(DB::PipelineExecutor::addJob(DB::ExecutingGraph::Node*)::$_0&)
<Fatal> BaseDaemon: 6.4. inlined from ../contrib/libcxx/include/functional:1608: std::__1::__function::__default_alloc_func<DB::PipelineExecutor::addJob(DB::ExecutingGraph::Node*)::$_0, void ()>::operator()()
<Fatal> BaseDaemon: 6. ../contrib/libcxx/include/functional:2089: void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::PipelineExecutor::addJob(DB::ExecutingGraph::Node*)::$_0, void ()> >(std::__1::__function::__policy_storage const*) @ 0x250f1f13 in /workspace/clickhouse
<Fatal> BaseDaemon: 7.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/functional:0: std::__1::__function::__policy_func<void ()>::operator()() const
<Fatal> BaseDaemon: 7.2. inlined from ../contrib/libcxx/include/functional:2560: std::__1::function<void ()>::operator()() const
<Fatal> BaseDaemon: 7. ../src/Processors/Executors/PipelineExecutor.cpp:597: DB::PipelineExecutor::executeStepImpl(unsigned long, unsigned long, std::__1::atomic<bool>*) @ 0x250eca9b in /workspace/clickhouse
<Fatal> BaseDaemon: 8.1. inlined from ./obj-x86_64-linux-gnu/../src/Processors/Executors/PipelineExecutor.cpp:0: DB::PipelineExecutor::executeSingleThread(unsigned long, unsigned long)
<Fatal> BaseDaemon: 8. ../src/Processors/Executors/PipelineExecutor.cpp:824: DB::PipelineExecutor::executeImpl(unsigned long) @ 0x250e77d3 in /workspace/clickhouse
<Fatal> BaseDaemon: 9.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:1245: std::__1::__compressed_pair_elem<DB::ExecutingGraph*, 0, false>::__get() const
<Fatal> BaseDaemon: 9.2. inlined from ../contrib/libcxx/include/memory:1334: std::__1::__compressed_pair<DB::ExecutingGraph*, std::__1::default_delete<DB::ExecutingGraph> >::first() const
<Fatal> BaseDaemon: 9.3. inlined from ../contrib/libcxx/include/memory:1627: std::__1::unique_ptr<DB::ExecutingGraph, std::__1::default_delete<DB::ExecutingGraph> >::operator->() const
<Fatal> BaseDaemon: 9. ../src/Processors/Executors/PipelineExecutor.cpp:410: DB::PipelineExecutor::execute(unsigned long) @ 0x250e6ac8 in /workspace/clickhouse
<Fatal> BaseDaemon: 10.1. inlined from ./obj-x86_64-linux-gnu/../src/Processors/Executors/PullingAsyncPipelineExecutor.cpp:81: DB::threadFunction(DB::PullingAsyncPipelineExecutor::Data&, std::__1::shared_ptr<DB::ThreadGroupStatus>, unsigned long)
<Fatal> BaseDaemon: 10.2. inlined from ../src/Processors/Executors/PullingAsyncPipelineExecutor.cpp:107: operator()
<Fatal> BaseDaemon: 10.3. inlined from ../contrib/libcxx/include/type_traits:3682: decltype(std::__1::forward<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&>(fp)()) std::__1::__invoke_constexpr<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&)
<Fatal> BaseDaemon: 10.4. inlined from ../contrib/libcxx/include/tuple:1415: decltype(auto) std::__1::__apply_tuple_impl<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&, std::__1::tuple<>&>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&, std::__1::tuple<>&, std::__1::__tuple_indices<>)
<Fatal> BaseDaemon: 10.5. inlined from ../contrib/libcxx/include/tuple:1424: decltype(auto) std::__1::apply<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&, std::__1::tuple<>&>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&, std::__1::tuple<>&)
<Fatal> BaseDaemon: 10.6. inlined from ../src/Common/ThreadPool.h:182: operator()
<Fatal> BaseDaemon: 10.7. inlined from ../contrib/libcxx/include/type_traits:3676: decltype(std::__1::forward<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(fp)()) std::__1::__invoke<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'()&>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)
<Fatal> BaseDaemon: 10.8. inlined from ../contrib/libcxx/include/__functional_base:348: void std::__1::__invoke_void_return_wrapper<void>::__call<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'()&>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&...)
<Fatal> BaseDaemon: 10.9. inlined from ../contrib/libcxx/include/functional:1608: std::__1::__function::__default_alloc_func<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'(), void ()>::operator()()
<Fatal> BaseDaemon: 10. ../contrib/libcxx/include/functional:2089: void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'(), void ()> >(std::__1::__function::__policy_storage const*) @ 0x2510c8c9 in /workspace/clickhouse
<Fatal> BaseDaemon: 11. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/functional:0: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xbdb7b8c in /workspace/clickhouse
<Fatal> BaseDaemon: 12.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:1655: std::__1::unique_ptr<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>, std::__1::default_delete<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()> > >::reset(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>*)
<Fatal> BaseDaemon: 12.2. inlined from ../contrib/libcxx/include/memory:1612: ~unique_ptr
<Fatal> BaseDaemon: 12. ../contrib/libcxx/include/thread:293: void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()> >(void*) @ 0xbdbf274 in /workspace/clickhouse
<Fatal> BaseDaemon: 13. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
<Fatal> BaseDaemon: 14. clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
<Fatal> BaseDaemon: Calculated checksum of the binary: 3A3FBC98213EAA6A2DBA9DF549079D64. There is no information about the reference checksum.
```

**Comment by alexey-milovidov** [comment]
> fuzz crash query

Does not reproduce. Maybe fixed.

**Pull request merged**